Chapter15 regions

HannaMichaelis opened the pull request:
## Summary

Contributes to Intellitect/EssentialCSharpManuscript#116

Keboo committed:
Changes from pdf review


Minor changes from pdf review


Chapter 13 first changes


Chapter 12 Regions through 12.6


Reviewed through 12.4


Changes from pdf review


Changes from pdf review


Minor changes from pdf review


Chapter 14 initial changes


chapter 14 code changes


Changes from pdf review


Minor changes from pdf review


Chapter 15 initial changes


chapter 13 review changes


adding ... to beginning and end


minor ch. 15 changes


preventing line rollovers ch. 14


final (hopefully) ch. 14 changes


Final ch. 15 changes


chapter 15 formatting changes


chapter 15 removing redundancies


Revert "Changes from pdf review"

This reverts commit e114c2cd40f0c3743dac45ed2264805f31b4a953.

Revert "Minor changes from pdf review"

This reverts commit 47d535e78b386baf3c5ec878a929ada981362996.

Revert "Chapter 13 first changes"

This reverts commit b211cf35264d0591d1f836a506f36598cecb3857.

BenjaminMichaelis commented on pp_region.snippet:
```diff
@@ -0,0 +1,28 @@
+			<Description>Essential C# Region</Description>
+			<Author>Microsoft Corporation</Author>
+			<SnippetTypes>
+				<SnippetType>Expansion</SnippetType>
```
What is this file for?

BenjaminMichaelis commented on src/Chapter14/Listing14.01.HeaterAndCoolerEventSubscriberImplementations.cs:
```diff
@@ -24,14 +26,15 @@ public void OnTemperatureChanged(float newTemperature)
         }
 
-        // Cooler is activated when ambient temperature is higher than this
+        // Cooler is activated when ambient temperature
```
```suggestion
        // Heater is activated when ambient temperature
```

BenjaminMichaelis commented on src/Chapter14/Listing14.01.HeaterAndCoolerEventSubscriberImplementations.cs:
```diff
@@ -24,14 +26,15 @@ public void OnTemperatureChanged(float newTemperature)
 
-        // Cooler is activated when ambient temperature is higher than this
+        // Cooler is activated when ambient temperature
+        // is higher than this
```
```suggestion
        // is lower than this
```
Since it's a heater would it activate when the temperature is lower than the goal?

BenjaminMichaelis commented on src/Chapter14/Listing14.02.DefiningTheEventPublsherThermostat.cs:
```diff
@@ -1,24 +1,14 @@
-        }
+        public float CurrentTemperature { get; set; }
+        #endregion INCLUDE
         private float _CurrentTemperature;
```
```suggestion
```
If we got rid of the fully qualified property I don't think we need this field

BenjaminMichaelis commented on src/Chapter14/Listing14.03.ConnectingThePublisherAndSubscribers.cs:
```diff
@@ -13,20 +12,23 @@ public static void Main()
             if (!int.TryParse(temperature, out int currentTemperature))
             {
-                Console.WriteLine($"'{temperature}' is not a valid integer.");
+                Console.WriteLine
+                    ($"'{temperature}' is not a valid integer.");
```
Is it intentional for this to be spread across two lines? I think it may make it difficult to read easily possibly?

HannaMichaelis replied:
The line is too long to fit on a single line

BenjaminMichaelis replied:
In the pdf? What occurs when it is too long? I think it may be clearer if in that case to maybe split halfway through the string, but I do not have strong opinions about this but maybe @Keboo or @MarkMichaelis have thoughts on this?

BenjaminMichaelis replied:
I am not fully clear yet though as to what it means it is too long to fit on a single line however.

HannaMichaelis replied:
It will rollover in the pdf, creating multiple lines, even in the middle of a word. It has to be split into multiple lines to avoid this. The line can definitely be split elsewhere but it does have to be split, otherwise, it is super unclear, especially with the indentations.

BenjaminMichaelis commented on src/Chapter14/Listing14.07.UsingthePlusEqualsAndMinusEqualsDelegateOperators.cs:
```diff
@@ -3,11 +3,13 @@
     using Listing14_01;
     using Listing14_05;
-
+    #region INCLUDE
+    #region EXCLUDE
     public class Program
     {
         public static void Main()
         {
+            #endregion EXCLUDE
```
Would it be simpler to just move the #region INCLUDE to just surround the inner section instead of the #exclude right next to an include?

BenjaminMichaelis commented on src/Chapter14/Listing14.07.UsingthePlusEqualsAndMinusEqualsDelegateOperators.cs:
```diff
@@ -16,18 +18,24 @@ public static void Main()
+            #region EXCLUDE
         }
     }
+    #endregion EXCLUDE
```
Would it be simpler to just move the #region INCLUDE to just surround the inner section instead of the #exclude right next to an include?

BenjaminMichaelis commented on src/Chapter14/Listing14.08.UsingThePlusAndMinusDelegateOperators.cs:
```diff
@@ -3,11 +3,13 @@
     using Listing14_01;
     using Listing14_05;
-
+    #region INCLUDE
```
Would it be simpler to just move the #region INCLUDE to just surround the inner section instead of the #exclude right next to an include?

BenjaminMichaelis commented on src/Chapter14/Listing14.08.UsingThePlusAndMinusDelegateOperators.cs:
```diff
@@ -22,12 +24,19 @@ public static void Main()
             delegate3 = (delegate3 - delegate2)!;
+            #endregion HIGHLIGHT
             delegate3(60);
+            #region EXCLUDE
         }
     }
+    #endregion EXCLUDE
+    #endregion INCLUDE
 }
```
Would it be simpler to just move the #region INCLUDE to just surround the inner section instead of the #exclude right next to an include?

BenjaminMichaelis commented on src/Chapter14/Listing14.09.OnTemperatureChangedThrowingAnException.cs:
```diff
@@ -13,27 +12,28 @@ public static void Main()
             if (!int.TryParse(temperature, out int currentTemperature))
             {
-                Console.WriteLine($"'{temperature}' is not a valid integer.");
+                Console.WriteLine
+                    ($"'{temperature}' is not a valid integer.");
```
I think splitting between two lines at this exact point in time may be confusing. Maybe something more similar to this if we are trying to shorten the lines?
```
                Console.WriteLine($"'{temperature}' 
                is not a valid integer.");
```
But I think maybe one line could work well here, thoughts?

HannaMichaelis replied:
I changed it to be like you said, it adds an extra "$" so just make sure it's what you want!

BenjaminMichaelis commented on src/Chapter14/Listing14.10.HandlingExceptionsFromSubscribers.cs:
```diff
@@ -19,9 +17,11 @@ public float CurrentTemperature
                 {
                     _CurrentTemperature = value;
-                    Action<float>? onTemperatureChange = OnTemperatureChange;
+                    Action<float>? 
+                        onTemperatureChange = OnTemperatureChange;
```
I think we want this to all be on one line

Keboo replied:
If we do need to split this across lines I would do it at the `=`
```C#
Action<float>? onTemperatureChange
    = OnTemperatureChange;
```

BenjaminMichaelis commented on src/Chapter14/Listing14.11.UsingTheAssignmentOperationRatherThanPlusEquals.cs:
```diff
@@ -14,24 +13,26 @@ public static void Main()
             if (!int.TryParse(temperature, out int currentTemperature))
             {
-                Console.WriteLine($"'{temperature}' is not a valid integer.");
+                Console.WriteLine
+                    ($"'{temperature}' is not a valid integer.");
```
Same comment here as earlier, maybe this should be made into one line? Thoughts anyone?

HannaMichaelis replied:
I changed it to be like you said, it adds an extra "$" so just make sure it's what you want!

BenjaminMichaelis commented on src/Chapter14/Listing14.18.CSharpConceptualEquivalentOfEventCILCode.cs:
```diff
@@ -33,19 +35,22 @@ public Thermostat(EventHandler<TemperatureArgs>? onTemperatureChange)
-        //        remove_OnTemperatureChange(value);
-        //    }
-        //}
+        #if ConceptulEquivalendCode
```
```suggestion
        #if ConceptualEquivalentCode
```
I think this is mispelled

Keboo replied:
I think "Equivalend" is also a misspelling

BenjaminMichaelis commented on src/Chapter14/Listing14.18.CSharpConceptualEquivalentOfEventCILCode.cs:
```diff
@@ -33,19 +35,22 @@ public Thermostat(EventHandler<TemperatureArgs>? onTemperatureChange)
+                remove_OnTemperatureChange(value);
+            }
+        } 
+        #endif // ConceptulEquivalendCode
```
```suggestion
        #endif // ConceptualEquivalentCode
```

BenjaminMichaelis commented on src/Chapter14/Listing14.19.CustomAddAndRemoveHandlers.cs:
```diff
@@ -11,27 +13,33 @@ public TemperatureArgs(float newTemperature)
             add
             {
-                _OnTemperatureChange = (TemperatureChangeHandler)System.Delegate.Combine(value, _OnTemperatureChange);
+                _OnTemperatureChange = 
+                    (EventHandler<TemperatureArgs>)System.Delegate.Combine
+                        (value, _OnTemperatureChange);
```
I think this might want to be made a single line

HannaMichaelis replied:
I changed it to be how it is written in the book, it's not a single line there

BenjaminMichaelis commented on src/Chapter14/Listing14.19.CustomAddAndRemoveHandlers.cs:
```diff
@@ -11,27 +13,33 @@ public TemperatureArgs(float newTemperature)
             remove
             {
                 _OnTemperatureChange = 
-                    (TemperatureChangeHandler?)System.Delegate.Remove(_OnTemperatureChange, value);
+                    (EventHandler<TemperatureArgs>?)System.Delegate.Remove
+                        (_OnTemperatureChange, value);
```
I think this may be easier to read in a single line

HannaMichaelis replied:
too long for a single line

HannaMichaelis replied:
I changed it to be how it is written in the book, it's not a single line there

BenjaminMichaelis commented on src/Chapter15/Listing15.19.AnInnerJoinUsingSystem.Linq.Enumerable.Join.cs:
```diff
@@ -26,12 +28,16 @@ public static void Main()
 
 
-            foreach ((int Id, string Name, string Title, Department Department) item in items)
+            foreach
+                ((int Id, string Name, string Title,
+                    Department Department) item in items)
```
I think this can be made into one line?
```suggestion
            foreach ((int Id, string Name, string Title, Department Department) item in items)
```

HannaMichaelis replied:
too long

BenjaminMichaelis commented on src/Chapter15/Listing15.20.AnotherInnerJoinWithSystem.Linq.Enumerable.Join.cs:
```diff
@@ -24,12 +27,14 @@ public static void Main()
 
 
-            foreach ((long Id, string Name, Employee Employee) item in items)
+            foreach 
+                ((long Id, string Name, Employee Employee) item in items)
```
I think this might be able to be one line?
```suggestion
            foreach ((long Id, string Name, Employee Employee) item in items)
```

HannaMichaelis replied:
Again, this was too long to fit on a single line

BenjaminMichaelis replied:
@MarkMichaelis thoughts?

BenjaminMichaelis commented on src/Chapter15/Listing15.22.CreatingAChildCollectionWithSystem.Linq.Enumerable.GroupJoin.cs:
```diff
@@ -1,18 +1,23 @@
             Employee[] employees = CorporateData.Employees;
 
-            IEnumerable<(long Id, string Name, IEnumerable<Employee> Employees)> items =
+            IEnumerable<
+                (long Id, string Name, IEnumerable<Employee> Employees)> 
+                    items =
```
This seems odd to be spread over 3 lines to me

HannaMichaelis replied:
Too long otherwise

BenjaminMichaelis commented on src/Chapter15/Listing15.22.CreatingAChildCollectionWithSystem.Linq.Enumerable.GroupJoin.cs:
```diff
@@ -23,14 +28,18 @@ public static void Main()
                     ));
 
-            foreach ((_, string name, IEnumerable<Employee> employeeCollection) in items)
+            foreach (
+                (_, string name, IEnumerable<Employee> employeeCollection) 
+                    in items)
```
I think this can be made into one line

HannaMichaelis replied:
Too long otherwise

HannaMichaelis replied:
I changed it to be how it is written in the book, it's not a single line there

BenjaminMichaelis commented on src/Chapter15/Listing15.27.ProjectionToAnAnonymousType.cs:
```diff
@@ -4,16 +4,18 @@ namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter15.Listing15_27
     using System.IO;
     using System.Linq;
-
+    #region INCLUDE
+    #region EXCLUDE
```
Would it be simpler to just move the #region INCLUDE to just surround the inner section instead of the #exclude right next to an include?

BenjaminMichaelis commented on src/Chapter15/Listing15.27.ProjectionToAnAnonymousType.cs:
```diff
@@ -36,4 +40,6 @@ private static void Print<T>(IEnumerable<T> items)
             }
         }
     }
+    #endregion EXCLUDE
+    #endregion INCLUDE
```
Would it be simpler to just move the #region INCLUDE to just surround the inner section instead of the #exclude right next to an include?

BenjaminMichaelis reviewed (changes requested):
This looks good overall! Just some minor changes and adjustments from my end

HannaMichaelis committed:
ch 14 & 15 removing redundancies

Keboo commented on src/Chapter14/Listing14.14.DeclaringAGenericDelegateType.cs:
```diff
@@ -4,7 +4,10 @@
     {
-        public delegate void EventHandler<TEventArgs>(object sender, TEventArgs e)
+        #region INCLUDE
+        public delegate void EventHandler<TEventArgs>(
+            object sender, TEventArgs e)
+            #endregion INCLUDE
             where TEventArgs : EventArgs;
```
I think we probably want to move the `#endregion` after the semicolon, otherwise it looks funny and is not valid C#.
```suggestion
        public delegate void EventHandler<TEventArgs>(
            object sender, TEventArgs e)
            where TEventArgs : EventArgs;
        #endregion INCLUDE
```

Keboo commented on src/Chapter15/Listing15.02.DictionaryInitializers.cs:
```diff
@@ -26,6 +28,7 @@ public static void Main()
                     {"Verbose", ConsoleColor.White}
                 };
 #endif
+            #endregion INCLUDE
```
Is this in the right place? I would have expected it after to the closing brace on line 34

HannaMichaelis replied:
That's where it is in the book but I'm sure it can be changed if it would make more sense elsewhere.

Keboo replied:
Nope works for me

Keboo commented on src/Chapter15/Listing15.17.OrderingWithSystem.Linq.Enumerable.OrderByAndThenBy.cs:
```diff
@@ -23,6 +25,9 @@ public static void Main()
             Print(items);
+
+            //...
+            #endregion INCLUDE
```
Is this in the right place? I would have expected it after the closing brace on line 40

HannaMichaelis committed:
Update src/Chapter14/Listing14.14.DeclaringAGenericDelegateType.cs

Co-authored-by: Kevin B <Keboo@users.noreply.github.com>

HannaMichaelis committed:
ch 14 changes after pull request

HannaMichaelis committed:
15.13 change

HannaMichaelis committed:
ch 14 & 15 single lines

HannaMichaelis committed:
Update src/Chapter15/Listing15.19.AnInnerJoinUsingSystem.Linq.Enumerable.Join.cs

Co-authored-by: Benjamin Michaelis <benjamin@michaelis.net>

HannaMichaelis committed:
Update src/Chapter15/Listing15.20.AnotherInnerJoinWithSystem.Linq.Enumerable.Join.cs

Co-authored-by: Benjamin Michaelis <benjamin@michaelis.net>

HannaMichaelis commented:
I've updated and changed everything that I know how. There are still errors in 14.19 and 15.13 that are blocking the checks but I don't know how to fix them.

BenjaminMichaelis commented:
Perfect! I will try to get those errors fixed and review the remaining comments

BenjaminMichaelis committed:
Fix errors

BenjaminMichaelis committed:
Apply suggestions from code review

BenjaminMichaelis committed:
Update src/Chapter14/Listing14.18.CSharpConceptualEquivalentOfEventCILCode.cs

BenjaminMichaelis committed:
Update src/Chapter14/Listing14.18.CSharpConceptualEquivalentOfEventCILCode.cs

BenjaminMichaelis committed:
Delete snippet file

BenjaminMichaelis commented:
Thanks @HannaMichaelis! I'll :shipit: now
